### DISCUSSION TITLE
关于pag模版混音的想法

### DISCUSSION THREAD

**Discussion by ridgeback1982** [Tencent/libpag#2558] [Ideas] [answered]
用AE导出一个装有两个视频（每个视频分别在一个子合成里）的pag模版。之所以把两个视频分别做成两个子合成（包含VideoComposition的PreComposeLayer），就是希望能在后期独立控制。结果发现音频audioBytes存在于顶层的VectorComposition里，没有在那两个VideoComposition里，说明AE插件在导出音频的时候做了混音。这样操作对于一般的应用自然没有问题，但是对于一个面向视频剪辑的应用，就限制的太死了。
如果能把每个视频的音频部分分别导出到各自VideoComposition里（VideoComposition也有audioBytes字段），那么后期做混音的时候会更加灵活（比如调整混音权重），这样岂不是更好？

**Comment by kevingpqi123** [comment, marked as answer]
对于 pag 而言，专注于特效渲染，内部没有音频处理模块，针对你所说的场景，pag 做为模版使用，在占位图中填充视频，占位图是 ImageLayer，不是 VideoComposition， 如果占位图中传入的视频内部有音频信息， 音频信息也不会在图层上，而是在视频中，需要业务方自己处理音频相关逻辑，如音频解码、音频混合和音频播放等。

VideoComposition 是 BMP 预合成，本质上是一个视频序列，如设计师在 AE 中使用了不能直接导出的特效、视频等，这个时候会采用 BMP 预合成的导出方式，VideoComposition 中的内容是不能编辑的，不能替换为其它视频，做模板占位图使用的是 ImageLayer，当然，设计师在 AE 中可以添加视频，如果没有勾选 BMP 预合成导出，导出的就是一个占位图，视频中的音频信息也不会导出，没有任何意义，这些音频信息视频编辑场景中也是用不到

pag 中 的 audioBytes 解决的是设计师想在 pag 文件中透传一些音频数据，如一个动画播放的过程中有一个固定的音效，如果不透传，就需要分两个文件提供给开发，非常不方便。这里的音频文件是不能编辑的，只能选择使用与不使用，因此放在 root Composition下面。

PAG 企业版中内置了音频混合和音频播放的功能，处理的也是占位图中传入的视频中的音频信息以及和 pag 文件自身的音频信息做混合，而不是将音频放在 VideoCompositon 中

**Reply by ridgeback1982** [reply]
明白了。pag中的audioBytes我可以理解为动画的固定音效，因为通常只有一个音效，所以就不存在混音的需求。企业版呢，可以在站位图里传音频，所以也就实现了混音。谢谢解释。
